### PR TITLE
refactor(text): rename `Spans` to `Line`

### PR DESCRIPTION
**Pull request by Eyesonjune18**
Because `Spans` conflicts with the plural `Span`s, and because the type represents a line of text, I've renamed it to `Line`. I will also be adding a deprecated type alias to avoid breaking existing codebases (haven't committed it yet).

### CODE REVIEW

**Comment by joshka** [comment]
This PR is a remake of #172. Some relevant comments from there (note it looks like the type alias / deprecated consideration for backward compatibility was made part of this change and the extra patch methods were removed), and that this is probably not a breaking change anymore:

From #163 (discussion on breaking changes):

1. Do we have a good reason for the breaking change?
  `Line` is a better description of what `Spans` purpose is
3. Have we considered non-breaking changes to fix the issue?
  Some suggestions in my review for approaching this (a `#[deprecated]` type alias and method.
5. Have we documented what the change is in detail before making it?
  This could be a little better in the commit message body
7. Are there other changes around the same area that would reduce the fatigue of responding to updates?
  There are a few PRs currently outstanding that might be worth shipping at the same time.
9. Have we documented how to fix the change in downstream code?
  This should just be a fairly easy rename: `find . -name '*.rs' -exec sed -i '' 's/Spans/Line/g' {} \;` and then check for any places that the rename conflicts (e.g. when using Canvas
11. Do we know how many dependents / migrations use the code being changed?
  Ballpark 3.5K files: https://github.com/search?q=%22use+tui%22+%22Spans%22+language%3Arust&type=code
12. Do we have a plan for communicating the change (and a timeline)?
  No
13. Do we need user feedback on this change?
  Probably
14. Should we wait a little for other things to settle before making this change?
  Maybe
15. What else?
  I don't know - what do you think?

**Comment by joshka** [comment]
Can you please rerun `cargo fmt` so the checks succeed?

I found that it's worth configuring my editor (VSCode) to automatically run rustfmt on save to avoid this problem, YMMV.

**Comment by Eyesonjune18** [comment]
> Can you please rerun `cargo fmt` so the checks succeed?
> 
> I found that it's worth configuring my editor (VSCode) to automatically run rustfmt on save to avoid this problem, YMMV.

Fixed.

**Review by joshka — APPROVED** [review]
LGTM - I'm generally for this going ahead as a change as it's non-breaking.

**Comment by Eyesonjune18** [comment]
Will there be any way to non-breakingly deprecate `Line.0`? I don't know if you can do that with fields.

**Comment by joshka** [comment]
> Will there be any way to non-breakingly deprecate `Line.0`? I don't know if you can do that with fields. It would be replaced with `Line.spans`.

Probably not if we went direct to this implementation. We could possibly leave `Spans` in the project, and instead modify every place that currently takes a `Spans` to instead take an `Into<Line>` and implement `From<Spans>` on `Line`. It could be worth attempting that to see if it works as an alternative to this change. It seems like that would have similar properties of letting us deprecate functionality in a way that notifies users that things will break soon with a compiler message.

I like that you're considering other changes that might be worth making at the same time on this. :)

**Comment by Eyesonjune18** [comment]
> I like that you're considering other changes that might be worth making at the same time on this. :)

Do you think this should be done in the same PR?

**Comment by sayanarijit** [comment]
Can we make Line.0 private and implement Into and From for Span list? Or maybe a public method spans()?

And keep Spans, not as a deprecated struct, but undocumented, maybe in a separate module spans.rs?

This way, as a lazy user, if I don't want to move to Line, I won't feel pressured. And Line will have a nicer, more flexible api.

**Comment by Eyesonjune18** [comment]
> Can we make Line.0 private and implement Into and From for Span list? Or maybe a public method spans()?
> 
> And keep Spans, not as a deprecated struct, but undocumented, maybe in a separate module spans.rs?
> 
> This way, as a lazy user, if I don't want to move to Line, I won't feel pressured. And Line will have a nicer, more flexible api.

If you can't be bothered to find-and-replace `.0` with `.spans` in your codebase, just don't update the crate. I'm fine with temporarily leaving `Spans` in the crate as a deprecated feature but I think we should then remove it on the next minor version release.

**Comment by Eyesonjune18** [comment]
> Let's actually discuss this whole compatibility thing in a meeting. That I believe is the best to go about this.

When are the meetings held?

**Comment by sayanarijit** [comment]
There was some discussion on having meetings on a scheduled time, but I was inactive for some time, so I need to check with others. Hope we're not in a hurry.

**Comment by Eyesonjune18** [comment]
> There was some discussion on having meetings on a scheduled time, but I was inactive for some time, so I need to check with others. Hope we're not in a hurry.

The main issue for me is that I'm needing new functionality to be developed for my own project (a shell) because as it stands it's going to be basically impossible to continue with the lack of features. I have a group of people working on it and I really don't want them to feel like the project has been abandoned.

**Comment by joshka** [comment]
> The main issue for me is that I'm needing new functionality to be developed for my own project (a shell) because as it stands it's going to be basically impossible to continue with the lack of features. I have a group of people working on it and I really don't want them to feel like the project has been abandoned.

Having specific needs to implement a feature is a really good counterbalance to the idea that breaking shouldn't happen. I'd definitely encourage you to share the blockers (because others might have them too, and we probably will mostly agree that breaking things to fix this is a good approach)

**Comment by Eyesonjune18** [comment]
> > The main issue for me is that I'm needing new functionality to be developed for my own project (a shell) because as it stands it's going to be basically impossible to continue with the lack of features. I have a group of people working on it and I really don't want them to feel like the project has been abandoned.
> 
> Having specific needs to implement a feature is a really good counterbalance to the idea that breaking shouldn't happen. I'd definitely encourage you to share the blockers (because others might have them too, and we probably will mostly agree that breaking things to fix this is a good approach)

Some particular pain points for me have been text wrapping (requires breaking), lack of documentation (does not require breaking), widget reusability/odd architecture requirement for generating frames (requires breaking), and cursor movement tracking (probably does not require breaking).

**Comment by sayanarijit** [comment]
> because others might have them too

Agree, and we actually need to assess whether this is truly the case. Otherwise I'm against breaking >300 crates to accomodate the need of one single crate.

**Comment by joshka** [comment]
I'd say that as this currently stands it can safely be merged, with the caveat that delaying merging this creates an opportunity to decide on whether to change to fields rather than a tuple.

As I don't think this is a fix that specifically blocks anyone from using the feature, we're not on a time crunch for shipping this, so let's experiment with whether that idea works and go from there.

TL;DR I've approved this PR in the current form, but my preference is to not merge it yet.

**Comment by joshka** [comment]
Actually, scratch that. I forgot that https://github.com/tui-rs-revival/ratatui/pull/149 makes the change to Spans already in order to add single line alignment. Perhaps @TimerErTim could weigh in on whether that change could be made backwards compatible (using the approach of leaving the current `Spans` as is and changing places that accept spans to accept e.g. `Into<Line>` (https://github.com/tui-rs-revival/ratatui/pull/175#issuecomment-1544915625)

**Comment by TimerErTim** [comment]
At first glance I see no reason why such approach would not be possible. 

On the contrary, we could even make #149 non-breaking by keeping the current `Spans` tuple struct as is and returning the new `Line` struct on `Spans::alignment`.

**Comment by Eyesonjune18** [comment]
> Agree, and we actually need to assess whether this is truly the case. Otherwise I'm against breaking >300 crates to accomodate the need of one single crate.

This is 300 crates that *could* use our crate, not that *do*, to my knowledge. Those aren't even remotely similar metrics in my opinion.

**Comment by joshka** [comment]
> This is 300 crates that _could_ use our crate, not that _do_, to my knowledge. Those aren't even remotely similar metrics in my opinion. People generally expect to need to do some modifications to their existing codebase when migrating dependencies.

Maybe let's keep that concern over on #163

**Comment by Eyesonjune18** [comment]
Can I get a second reviewer on this?

**Comment by TimerErTim** [comment]
Should we take the approach of leaving `Spans` as is and refactoring for ´impl Into<Line>`?

If yes, that change would enable a smooth transition between working and breaking by marking `Spans` as deprecated. I would then (if there is no blocker) merge the PR quickly so other PRs can build upon the change/addition and make use of the new `Line` struct.

**Comment by mindoodoo** [comment]
Yeah either that or a `.spans()` method as @sayanarijit mentionned earlier, but one of the two would be good I think.

Other than that, lgtm @Eyesonjune18

**Comment by Eyesonjune18** [comment]
> Yeah either that or a `.spans()` method as @sayanarijit mentionned earlier, but one of the two would be good I think.
> 
> Other than that, lgtm @Eyesonjune18

I'm confused as to what you mean by those two suggestions. Can you give me an example?

**Comment by joshka** [comment]
Closing as #178 is merged - thank you for the robust conversation and initial implementation on this one @Eyesonjune18

**Pull request closed without merging**
